### PR TITLE
PERF: unstack

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

Avoids expensive get_indexer calls.

```
from asv_bench.benchmarks.reshape import *

self = Unstack()
self.setup("category")

%timeit self.time_full_product("category")
58.8 ms ± 1.52 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)  # <- master
31.6 ms ± 1.98 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)  # <- PR
```

### CODE REVIEW

**Comment by jreback** [comment]
does this allow us to get rid of some cython code (that i wrote a long time ago) to do unstacking?

**Comment by jbrockmendel** [comment]
> does this allow us to get rid of some cython code (that i wrote a long time ago) to do unstacking?

no

File: pandas/core/reshape/reshape.py [hunk 1]
```diff
@@ -287,7 +287,7 @@ def get_new_values(self, values, fill_value=None):
 
         return new_values, new_mask
 
-    def get_new_columns(self, value_columns):
+    def get_new_columns(self, value_columns: Index | None):
```
Comment (jreback) [inline pandas/core/reshape/reshape.py:290]: is this still called?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jbrockmendel** [comment]
Yes, just fewer times

**Review by jreback — APPROVED** [review]

**Pull request merged**